casper/src/main/rholang/BasicWallet.rho

```diff
@@ -0,0 +1,65 @@
+//scalapackage coop.rchain.rholang.wallet
+
+//assumes standard purse API from MakeMint
+contract @"BasicWallet"(@purse, @algorithm, @pk, return) = {
+  new cryptoVerifyCh in {
+    match algorithm {
+      "ed25519" => {
+        cryptoVerifyCh!(["ed25519Verify"])
+      }
+      _ => {
+        cryptoVerifyCh!([])
+      }
+    } |
+    for(@maybeCryptoVerify <- cryptoVerifyCh) {
+      match maybeCryptoVerify {
+        []             => { return!([]) }
+        [cryptoVerify] => {
+          new basicWallet, verify, nonceCh in {
+            nonceCh!(-1) |
+            contract verify(@nonce, @amount, @retCh, @sig, return) = {
+              for(@prevNonce <- nonceCh) {
+                match nonce == (prevNonce + 1) {
+                  true => { 
+                    new result in {
+                      @cryptoVerify!([nonce, amount, retCh].toByteArray(), sig.hexToBytes(), pk.hexToBytes(), *result) |
+                      for(@r <- result) {
+                        match r {
+                          true => { nonceCh!(nonce) | return!(true) }
+                          _    => { nonceCh!(prevNonce) | return!(false) }
+                        }
+                      }
+                    }
+                  }
+                  _ => { nonceCh!(prevNonce) | return!(false) }
+                }
+              }
+            } |
+            contract @[*basicWallet, "getNonce"](return) = {
+              for(@nonce <- nonceCh) {
+                nonceCh!(nonce) | return!(nonce)
+              }
+            } |
+            contract @[*basicWallet, "getBalance"](return) = {
```

Comment:
Have you considered how this "method dispatch" idiom interacts with ocap patterns such as [revocable forwarder][1]?

In another idiom I've seen, the `"getBalance"` method name is sent along with the args. It's more clear to me how that would compose with revocable forwarder.

I started a [forum discussion on method dispatch idioms][2].


[1]: https://developer.rchain.coop/tutorial/#revocation
[2]: https://forum.rchain.coop/developer/rholang/ocap-happy-method-dispatch-idioms-in-rholang/

Reply:
I don't think this idiom is less composable than sending string method names explicitly. In this model capabilities are  given to a name based on the contracts contained at "compound names" you can form with it. The fully featured object has contracts at all the compound names, but you can create a new name with fewer capabilities easily:
```
contract @"createRevocableForwarder"(baseName, allowedMethods, return) = {
  new isAlive, forwarder, addMethod, kill in {
    isAlive!(true) |
    contract addMethod(@method, success) = {
      //there is a minor problem here that you need to handle different
     //number of arguments somehow
      for(@args <- @[*forwarder, method]; _ <- isAlive) {
        @[*baseName, method]!(args) | 
        success!(true) |
        isAlive!(true) | 
        addMethod!(method, Nil)
      }
    } |
    contract kill(success) = {
      for(_ <- isAlive) {
        //Do not return any data to the `isAlive` channel, so the 
       //join above will no longer trigger.
        success!(true)
      }
    } |
    //initialize
    new methodsAdded in {
      @["LinkedList", "foreach"]!(allowedMethods, *addMethod, *methodsAdded) |
      for(_ <- methodsAdded) {
        return!([*forwarder, *kill])
      }
    }
  }
}
```
The contract above creates a revocable forwarder with any number of capabilities, which are specified at construction of the forwarder. There is a minor hiccup in that if the methods you wanted to forward took multiple arguments then `args` would need to be a list and you would need to deconstruct that list and pass its elements on to the underlying method.

Reply:
Thanks for spelling that out in detail.

It seems a little awkward, but maybe that's mostly a matter of what I'm familiar with.